### ISSUE TITLE
Private https sharing folder

### ISSUE THREAD

**Issue by MagnaCapax** [MagnaCapax/PMSS#1]
Password protected + higher rate limits than the public folder for sharing.

User request.

Requires some sort of pw management for separate htpasswd file, otherwise very quick to do

**Comment by MagnaCapax** [comment]
High-level ideation (no implementation yet):

- Expose a share endpoint like /fetch-user/file.php?id=<token> or /share/<token>.
- Use an opaque random token (>=128-bit) mapped server-side to file path + owner + expiry; avoid raw SHA256-of-path so tokens cannot be guessed.
- If stateless tokens are preferred, use an HMAC-signed payload (path id + expiry + scope) with a per-host secret; keep a denylist or rotate secrets for revocation.
- Enforce TTL, optional max-downloads/password, and audit logging (log token hash only).
- Stream the file after access checks; avoid leaking usernames/paths in URLs.

— Sampsa Pellervoinen 🌱

**Comment by MagnaCapax** [comment]
## Investigation Report

**Issue:** #1 — Private https sharing folder
**Investigated:** 2026-05-31
**Status:** PARTIAL — most of the building blocks already exist in the current codebase

### Summary
The request is roughly 80% already delivered by infrastructure that has landed since this issue was filed in 2023. The authenticated `/user-<user>/` path is already a password-protected web area that serves arbitrary `www/` files, and nginx already gives it a *higher* rate limit (100 MB/s) than the public folder (32 MB/s). The only genuinely missing piece is a **separate, shareable credential** distinct from the customer's master login — and that belongs in the password-management consolidation already tracked by #482, not in a standalone feature.

### What The Issue Asks For
> "Password protected + higher rate limits than the public folder for sharing. … Requires some sort of pw management for separate htpasswd file, otherwise very quick to do."

Decomposed: (1) a password-protected sharing folder, (2) with a higher rate limit than the public folder, (3) using a *separate* htpasswd credential so the share password isn't the master login.

Note: the two ideation comments drift toward a much larger design — a tokenized per-file share endpoint (`/share/<token>`, HMAC/opaque tokens, TTL, max-downloads, audit log). That is a different and far bigger feature than the htpasswd-folder the issue body actually requests. KISS/YAGNI: this report evaluates the simple folder, not the token service.

### Implementation Status
Already present today:

- **Public sharing folder** — `www/public/` served unauthenticated at `/public/` and `/public-<user>/` (`etc/seedbox/config/template.lighttpd:17,80-81`). Intentionally public by design (KB #71).
- **htpasswd auth** — `auth.backend="htpasswd"`, userfile `~/.lighttpd/.htpasswd`, protecting `/user-<user>/` and `/webdav-<user>/` (`template.lighttpd:58-74`).
- **An authenticated sharing area** — `/user-<user>/` serves arbitrary files from `www/` behind the user's password. nginx explicitly labels it the "Private area" and rate-limits it at **`limit_rate 102400k` (100 MB/s)** vs the public folder's **`limit_rate 32768k` (32 MB/s)** (`etc/seedbox/config/template.nginx-user:3,12-14,18,38-40`). So requirement (2), "higher rate than public," is already satisfied for the authenticated path.
- **Per-path rate limiting** lives in nginx. lighttpd's own `connection/server.kbytes-per-second` (204800/409600) are clamped to `0` (= unlimited) because both exceed the 65535 cap in `pmssClampLighttpdBandwidthLimits()` (`scripts/lib/lighttpd/configRender.php:11-21`). Per-path bandwidth differences should therefore be made in the nginx template (consistent with ADR-0009).
- **A per-user config-injection mechanism** — the template includes `~/.lighttpd/custom.d/*.conf` (`template.lighttpd:55-56`), and PMSS already writes *managed* fragments there (`pmss-deluge.conf`, `pmss-invidious.conf`, rclone, qbittorrent) via `pmssLighttpdWriteManagedProxyFragment()` (`scripts/lib/lighttpd/userConfigApply.php:48-83`), with the watchdog auto-restarting lighttpd on change. This is the natural home for a private-folder alias+auth fragment — written as managed code, not a manual drop-in (manual `custom.d` edits get overwritten on regeneration).
- **htpasswd write primitives** — `scripts/lib/lighttpd/htpasswd.php`.

### What's Missing
Exactly one thing: **a credential that is separate from the master login.** `changePw.php` manages a single credential — it sets the *same* password for the Unix account and the one `.htpasswd` entry (entry name = username), `scripts/changePw.php:83-115`. There is no second/share/guest credential mechanism anywhere in the tree. That single gap is what would let a customer grant folder access without handing out their master password.

A simplification worth flagging: the "separate htpasswd **file**" framing may be unnecessary. Adding a second entry (e.g. `share-<user>`) to the *existing* `.htpasswd` and scoping the private path with `require => "user=share-<user>"` achieves the same isolation without a second file or lighttpd named-backend syntax.

### Product & Privacy Rationale
- **Privacy** — a separate share credential is a genuine win (share without exposing the master login). But the public folder already covers non-sensitive sharing, so the marginal benefit is small.
- **Efficiency / reliability** — building a standalone credential subsystem + UI for a capability that is 80% present adds moving parts and failure surface (credential provisioning is already fragile — see #463) for little gain. Folding the remaining 20% into existing work serves both.
- **Investigate then act** — the "very quick to do" note was true against the 2023 architecture; against the 2026 architecture the quick parts are mostly already done.

### Devil's Advocate
Why NOT build this:
1. The authenticated `/user-<user>/` path already provides password-protected, higher-than-public-rate sharing of any file in `www/`.
2. The only delta (separate credential) overlaps directly with #482's password-management refactor.
3. No recurring demand for the password-protected variant has surfaced; observed sharing demand targets the public folder, which works.
4. The "separate file" sub-requirement is itself avoidable (second entry in the existing file).

Rebuttal (why the underlying need is still valid): sharing the master login to grant folder access is a real privacy/security compromise. The need is legitimate — but it is small and best delivered as an extension of #482 plus a thin path-config layer, not a new subsystem.

### Anti-Pattern Hunt
- **REGEX_FOR_SEMANTICS: PASS** — no semantic-matching-by-regex proposed. (The existing `configRender.php` clamp regex is pre-existing code, not a recommendation.)
- **PREMATURE_OPTIMIZATION: PASS (with flag)** — the *issue body* is not premature optimization, but the *comment-thread token design* (TTL/HMAC/audit) is gold-plating vs a demonstrated need; recommendation steers away from it.
- **FALSE_COMPLETION_CLAIM: PASS** — status is PARTIAL/~80%, every claim tied to a cited file:line; nothing claimed "done."
- **OPERATOR_HANDOFF_ANTIPATTERN: PASS** — recommendation is a concrete engineering decision, not a punt to the maintainer.
- **FABRICATED_AUTHORITY_OR_PATH: PASS** — every cited file, line range, and ADR was read from source during this investigation.
- **FAIL_OPEN_SAFETY_BEHAVIOR: PASS (noted)** — lighttpd's bandwidth clamp renders limits >65535 to `0`=unlimited, i.e. it fails *open* to unlimited; this is by design, with nginx as the enforcing layer. Any per-path private limit must be set in nginx, or it will be clamped away.
- **DRY_RUN_THEATER: PASS** — read-only investigation; no simulated execution claimed.
- **INVESTIGATION_LOOP_THEATER: PASS** — this is the first completed deep-dive on this issue and it reaches a decision rather than re-queuing.

### Related Demand & Data
No recurring customer demand for a *password-protected* sharing folder surfaced in support history; the observed sharing demand consistently targets the existing public folder (which works as designed). One recent contact was a customer confused between the authenticated `/user-` path and the public `/public-` path — a docs/UX gap, not a missing feature. The original 2023 "User request" remains the only recorded demand signal, unrenewed since. Low frequency = low priority signal.

### Cross-Issue Analysis
- **#482** (changePw refactor — "syncs all per-account passwords") — **MERGE**: the share-credential half of this issue is one more credential for #482 to manage. Natural owner.
- **#52** (Invidious public+private) — **DEPENDS**: same `/public-` vs `/user-` path split and the same managed `custom.d` fragment pattern.
- **#396** (Nextcloud in `public/`) — **DEPENDS**: same `www/public/` surface and lighttpd-PHP concern.
- **#4** (libreSpeed public + abuse limits) — **CONFLICTS in direction**: #4 wants *lower*/abuse-capped public limits, this wants *higher* authenticated limits; both touch the same per-path nginx rate-limit pattern, so design them together.
- **#463** (closed — addUser failed to create `.htpasswd`) — any second credential file inherits the same provisioning-gap risk; reuse the same creation path.

### Dependencies & Blockers
- The credential half should land on or after **#482** (shared password-management surface).
- The per-path nginx limit design should be coordinated with **#4** to avoid contradictory limits.
- Implementation detail to verify on a test host before coding: whether the deployed lighttpd honors the chosen per-path approach, and that a private fragment survives config regeneration (it must be managed code, not a manual `custom.d` edit).

### Reuse, Refactor & Test Strategy
- **Existing reuse points** — managed `custom.d` fragment writer (`userConfigApply.php` / `pmssLighttpdWriteManagedProxyFragment`); htpasswd primitives (`scripts/lib/lighttpd/htpasswd.php`); the nginx `/user-` location block as a copy-template for a `/private-` block; `changePw.php` as the credential model.
- **Refactor opportunities** — generalize the managed-fragment writer to also emit an auth+alias fragment (not only reverse-proxy fragments); extend the htpasswd/changePw path to manage a *named* second credential (this is the #482 merge).
- **Test categories** — unit (render a `/private-` fragment + second-entry htpasswd for a valid/invalid username); integration (lighttpd picks up the fragment and challenges `/private-`: 401 without the share credential, 200 with it); regression (config regeneration does not drop the managed fragment; `$HTTP[...]` quoting in `custom.d` per the #428 fix); boundary (username validation in render); failure injection (missing share-htpasswd behaves safely — mirror the #463 provisioning gap).
- **Minimum-code-for-correctness** — a second entry in the existing `.htpasswd` + one managed `custom.d` auth/alias fragment + one nginx `/private-` location, with the credential managed via #482. No token store, no new file format, no new daemon.
- **Success signal** — measurable: customers actually using a `/private-` share, or a drop in "how do I share privately" contacts. If neither materializes, the feature is decorative — which is precisely why the recommendation is to not build it standalone now.

### Architectural Minimization
- **Question the requirement:** "separate htpasswd **file**" is challenged — a second username entry in the existing `.htpasswd`, scoped via `require => "user=share-<user>"`, gives the same credential isolation without a second file or named-backend syntax. The underlying constraint (a share password that is not the master login) is real; the "separate file" mechanism is not.
- **Delete:** (1) don't add a `/private-` path — direct users to the existing authenticated `/user-` path (already password-protected, 100 MB/s); insufficient only because it requires sharing the master login. (2) don't add a separate htpasswd *file* — second entry in the existing file (recommended simplification).
- **Merge:** (1) credential management → **#482**; (2) path/rate → fold into the existing nginx per-location pattern + a managed `custom.d` fragment, not a new subsystem.
- **Refactor:** generalize the managed-fragment writer to emit auth/alias fragments; generalize the single-entry htpasswd writer to a named-credential writer (the #482 work).
- **Null:** if not built, customers share their master login for authenticated sharing, or use the public folder for non-sensitive files — a real but minor privacy cost that has not been hit in practice.
- **Conclusion:** a MERGE path wins, so no standalone minimal design is proposed.

### ADR Adherence
- **ADHERES: ADR-0009** (nginx lightweight reverse proxy; per-user lighttpd owns auth) — any private-folder auth must be in lighttpd, the rate-limit location in nginx; the recommendation does exactly this.
- **ADHERES: ADR-0016** (customer-facing PHP tree separation) — any user-facing "set share password" page must live under `etc/skel/www/`, with the operator-side script under `/scripts/`.
- **Context: ADR-0010** (fail2ban not enabled fleet-wide) — brute-force protection for a public-facing share password remains the user's responsibility, not a new fleet daemon.

### Recommendation
**SHOULD-NOT-IMPLEMENT** (as a standalone feature).

Reasoning: ~80% of the request is already delivered (authenticated `/user-` sharing at a higher rate than the public folder, htpasswd auth, managed `custom.d` injection, htpasswd primitives). The single genuine gap — a separate shareable credential — belongs in **#482**'s password-management consolidation, and the thin path/rate layer is a few lines mirroring the existing nginx `/user-` block plus a managed lighttpd fragment. With no recurring demand in 2.5 years, building a standalone credential subsystem + UI is redundant work that adds failure surface. The underlying privacy need is valid; the right delivery is *fold the credential into #482 + add the path layer on demand*, not a new feature. Maintainer may reprioritize if demand appears — the minimal build path is documented above.

### Self-Challenge (disproof attempts)
- **Strongest finding 1** — "higher rate than public is already met by the authenticated path." Disproof attempt: re-read the nginx `/user-` and `/public-` blocks in full. Result: re-confirmed (`/user-` = 100 MB/s, explicitly the "Private area"; `/public-` = 32 MB/s). Held.
- **Strongest finding 2** — "only one credential is managed; no share-credential mechanism exists." Disproof attempt: searched the whole script tree and the customer PHP tree for any second/share/guest credential. Result: none found. Held.
- **Least-evidenced finding** — "no demand for the password-protected variant." Disproof attempt: direct search of support history for password-protected/private-sharing requests. Result: only generic password-hygiene boilerplate, no feature demand. Held, calibrated to: no demand in the recent corpus; the 2023 "User request" is the sole recorded signal.

### Closure
Leaving this open for maintainer prioritization. The recommendation is SHOULD-NOT-IMPLEMENT for redundancy/low-demand reasons (not supersession), and #482 owns only the credential half — so this is a maintainer priority call on the oldest open issue, not an automatic close.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Re-verification (2026-06-02)

Re-checked the 2026-05-31 investigation above against current `main` (HEAD `6d0dd47b`). The load-bearing facts still hold and the recommendation is unchanged — this is a verification pass, not a new design.

- **"Higher rate than the public folder" is already delivered for the authenticated path.** `etc/seedbox/config/template.nginx-user` still rate-limits the `/user-` "Private area" at `limit_rate 102400k` (≈100 MB/s) vs `/public-` at `32768k` (≈32 MB/s).
- **The one genuine gap remains** a credential separate from the master login: `scripts/changePw.php` still manages a single credential (one `.htpasswd` entry = the Unix login); no second/share/guest credential mechanism exists in the tree.
- **Only one commit touched the relevant files** since the report — `9b94fbc1` (`refactor(dry): json output`), a 2-line change to `changePw.php` that does not alter credential or rate-limit behavior. The lighttpd bandwidth clamp (`pmssClampLighttpdBandwidthLimits`, values >65535 → `0`/unlimited) is unchanged, so per-path limits belong in nginx (ADR-0009).
- No new demand for the password-protected variant has surfaced.

**Disposition unchanged: SHOULD-NOT-IMPLEMENT as a standalone feature.** ~80% is already present; the remaining piece (a separate shareable credential) is best delivered as part of the #482 password-management work plus a few lines mirroring the existing nginx `/user-` block — not a new subsystem. This is a maintainer prioritization call on the oldest open issue; the minimal build path is documented in the 2026-05-31 report.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Re-verification (2026-06-03)

Re-checked the investigation above against current `main` (HEAD `8bb85ef`). The load-bearing facts are unchanged and the recommendation stands — this is a verification refresh, not a new design.

- The authenticated private area already serves password-protected files at a **higher rate limit than the public folder**, so "higher rate than public" is already delivered for the authenticated path.
- The single genuine gap is still a credential **separate from the master login**: password management sets one credential per user, and no separate share/guest credential exists yet. That piece is best delivered as part of the password-management consolidation in #482 plus a thin per-path config layer — not a standalone credential subsystem.
- The only change to any relevant file since the last check is an unrelated lighttpd error-page tweak; nothing touches auth, the public/private split, or the rate limits.

**Disposition unchanged: SHOULD-NOT-IMPLEMENT as a standalone feature** (~80% already present). This is a maintainer prioritization call on the oldest open issue; the minimal build path is documented in the 2026-05-31 report above.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Adversarial Review

**Issue:** #1 — Private https sharing folder
**Reviewed:** 2026-06-08
**Verdict:** SHOULD-NOT-IMPLEMENT (as a standalone feature)

This is the adversarial pre-implementation pass. It concurs with the investigation already on this thread (2026-05-31, re-verified 06-02 / 06-03). Rather than re-derive that architecture trace, this pass independently re-verified the load-bearing facts against current `main` and adds the missing disposition action.

### Independent re-verification (current `main`, HEAD `b1644122`)
Prior reviews are treated as leads, not ground truth; the anchors were re-read from source this pass and all hold:

- **"Higher rate than the public folder" is already delivered for the authenticated path.** `etc/seedbox/config/template.nginx-user` rate-limits the `/user-` "Private area" at `limit_rate 102400k` (≈100 MB/s, line 13) versus `/public-` at `32768k` (≈32 MB/s, line 39).
- **The one genuine gap remains a credential separate from the master login.** `scripts/changePw.php` still drives `passwd` for the Unix account and rewrites a single `~/.lighttpd/.htpasswd` entry (lines 6–8, 84–101); no second / share / guest credential mechanism exists anywhere in the tree.
- **Zero state delta on the load-bearing files since the last review (06-03).** `git log` since `8bb85ef` touches none of `template.nginx-user`, `template.lighttpd`, `changePw.php`, or `scripts/lib/lighttpd/*`. The single nearby nginx change (`569ad48f`, Refs #641) edits `template.nginx-site-default` (public testfile concurrency) — it does not touch the auth or sharing rate-limit surface.

Verdict mechanically HELD.

### Adversarial Findings
- **Worst case if built as the comment-thread design (`/share/<token>`, HMAC/opaque tokens, TTL, max-downloads, audit log):** a new public-facing, user-reachable token endpoint → token-mapping or expiry bug exposes another user's path → silent cross-account read → discovered late via audit-log review, not at request time → credential/secret rotation and per-token revocation cleanup. That is a credential subsystem with its own failure surface, built for a request the issue body never made.
- **Inversion (guarantee maximum damage):** ship a stateless hash-of-path token with no per-host secret and no revocation — exactly the guessable-token failure the thread's own ideation warned against. The proposed simple form (second htpasswd entry) is far from that failure mode; the token form is one design slip away from it.
- **Regret in 6 months:** a standalone share-credential subsystem duplicates the credential management that #482 ("syncs all per-account passwords") is already chartered to own — two code paths writing user credentials, drifting apart.
- **Removal inversion:** if this feature already existed as a standalone subsystem, we would likely fold it into #482 rather than fight to keep it separate — which is the argument for not building it standalone now.
- **Dumbest implied requirement:** "separate htpasswd **file**." A second username entry (e.g. `share-<user>`) in the existing `.htpasswd`, scoped with `require => "user=share-<user>"`, gives identical credential isolation with no second file and no named-backend syntax.
- **Deletion/merge opportunity:** the request decomposes into (1) credential half → MERGE into #482, and (2) a thin path/rate layer mirroring the existing nginx `/user-` block plus a PMSS-managed `custom.d` fragment. No new subsystem.

### Caller Analysis
Not applicable: this is a feature-**add** request and the verdict is not to add it. No existing functionality is removed, replaced, or weakened, so the removal/caller-enumeration gate does not trigger.

### Merit Check (Tier A)
1. **Concrete present demand:** Theoretical — no present demand. The 2023 "User request" line is the sole recorded signal, unrenewed in ~2.5 years; observed sharing demand consistently targets the existing public folder.
2. **Function:** Adds a password-protected web folder, served at a higher rate limit than the public folder, gated by a credential distinct from the customer's master login.
3. **Success metric:** Within 30 days of shipping, measurable use of the new private-share path (request logs) or a drop in "how do I share privately" contacts. Neither signal is currently present, so the metric would most likely read zero — itself a SHOULD-NOT signal.
4. **Catastrophic if implemented:** Standalone credential provisioning is added → it inherits the known provisioning-gap class (cf. #463, where account creation failed to create `.htpasswd`) → a share credential silently fails to provision → the customer's private path is either unreachable or, worse, falls open depending on the fragment's default → discovered only when a customer reports it. Two credential write paths (this + #482) also drift over time.
5. **Catastrophic if NOT implemented:** A customer wanting authenticated sharing must either share their master login (a real but minor MISSION-privacy cost) or use the public folder for non-sensitive files. No customer has hit this in the logged corpus. Theoretical, low-severity — the asymmetry versus #4 is the verdict signal.
6. **Deletion alternative:** Direct users to the existing authenticated `/user-` path (already password-protected at ≈100 MB/s); add no `/private-` subsystem. The only true gap (a non-master credential) folds into #482.
7. **Cost:** (a) Build — a standalone subsystem + a customer-facing "set share password" page is materially more than the issue's "very quick to do" estimate; the minimal merge path is a few lines. (b) Maintenance/yr — a second credential write path requires ongoing sync-correctness review against `changePw.php`/#482 each time credential handling changes. (c) Blast radius if broken — per-user web auth on affected hosts; a credential-default mistake is a privacy exposure, not just an outage. (d) Undo — clean `git revert` for the config layer, but a customer-facing credential UI is a softer one-way door once customers rely on it.
8. **Dependency blast:** A second `.htpasswd` entry / new path interacts with config regeneration (managed `custom.d` fragments are overwritten on regen — a manual drop-in would be silently lost), the lighttpd watchdog restart, and #482's credential consolidation. These must be designed together, not in isolation.
9. **Requirements challenge:** The issue assumes a *new* protected sharing area is needed; the assumed baseline is already false — the authenticated `/user-` path is that area, and the residual requirement is narrowly "a credential that isn't the master login," not a folder/subsystem.

### Anti-Pattern Hunt
- **REGEX_FOR_SEMANTICS:** PASS — no semantic-by-regex logic proposed.
- **PREMATURE_OPTIMIZATION:** PASS (issue body) / FLAG (the thread's token-service ideation is gold-plating versus a demonstrated need; steer to the simple form if ever built).
- **OPERATOR_HANDOFF_ANTIPATTERN:** PASS — the engineering decision is resolved here (don't build standalone; fold credential → #482; thin path layer on demand). Only the product-priority call on a 3-year-old, near-zero-demand request is routed to the maintainer, with a concrete recommendation.
- **FALSE_COMPLETION_CLAIM:** PASS — status is PARTIAL/~80%; nothing is claimed done.
- **FABRICATED_AUTHORITY_OR_PATH:** PASS — every cited file/line/issue was read from source this pass.
- **FAIL_OPEN_SAFETY_BEHAVIOR:** PASS (noted) — lighttpd's bandwidth clamp renders values >65535 to `0` (unlimited), i.e. fails open by design with nginx as the enforcing layer; any per-path private limit must live in the nginx template (ADR-0009), or it is clamped away.
- **DRY_RUN_THEATER:** PASS — read-only investigation.
- **INVESTIGATION_LOOP_THEATER:** PASS — this pass does not re-derive the converged analysis; it re-verifies the anchors at the current HEAD and adds the disposition action (label) that prior passes named but could not apply.

### Recommendations
- **For the maintainer (the one open decision):** either close #1 as covered-by-direction, or explicitly fold its credential half into **#482** and the path/rate half into the existing nginx `/user-` pattern. The underlying privacy need is valid but small; this is a prioritization call on the oldest open issue, not an engineering gap.
- **If ever built (minimal path, for the implementer):** a second entry in the existing `.htpasswd` + one PMSS-managed `custom.d` auth/alias fragment + one nginx `/private-` location copied from the `/user-` block. No token store, no new file format, no new daemon. Auth in lighttpd, rate-limit in nginx (ADR-0009); customer-facing page under `etc/skel/www/`, operator script under `/scripts/` (ADR-0016).
- **Forward condition:** further review is not warranted unless one of these changes — any of the load-bearing files above is modified, #482 lands a credential mechanism, or fresh user demand for the password-protected variant appears.

### Worth-Implementing Analysis
- **Cost:** medium-high as a standalone subsystem; low as the #482 merge + thin path layer.
- **Benefit:** small and unproven — a real but minor privacy improvement over the existing authenticated path, with no recurring demand in ~2.5 years.
- **Verdict rationale:** ~80% is already delivered; the residual 20% (a non-master credential) has a natural owner in #482, and the deletion/merge alternative exists. Building a standalone subsystem adds failure surface for little measurable gain.
- **Alternative:** fold the credential into #482; add the path/rate layer only if demand materialises.

### Decomposition Targets
None proposed by this pass. The relevant reuse/refactor points (generalising the managed-`custom.d` fragment writer to emit auth/alias fragments, and generalising single-entry htpasswd writing to a named-credential writer) are the #482 work and are documented in the 2026-05-31 investigation on this thread; raising separate refactor issues would duplicate that.

— Väinämöinen
